### DISCUSSION TITLE
Misleading openapi_version parameter in get_openapi

### DISCUSSION THREAD

**Discussion by tringwald** [fastapi/fastapi#10198] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Optional

import fastapi.openapi
import pydantic


class DummyRequest(pydantic.BaseModel):
    value: Optional[int] = None


app = fastapi.FastAPI()


@app.post("/test")
def get_dummy(payload: DummyRequest) -> None:
    pass


if __name__ == '__main__':
    openapi_spec = fastapi.openapi.utils.get_openapi(title='title', version='1.0.0', openapi_version='3.1.0', routes=app.routes)
    print(openapi_spec)
```

### Description

Recently, I wanted to export the OpenAPI Spec of my fastapi app in order to send it to a third party for some automated processing. For this, I used the `fastapi.openapi.utils.get_openapi` function (see dummy code above). I could not use `app.openapi()` as I just wanted to export a part of the API.

Now, the code above works perfectly fine and produces a valid OAS. Unfortunately, the third party turned out to only accept OAS 3.0.x. I thus modified above code, setting `openapi_version='3.0.0'`. However, contrary to my expectations, this did not produce a valid OAS 3.0.0 dict. It did in fact just produce the same dict again, but this time with the `openapi` key set to `3.0.0`. 

According to the [code](https://github.com/tiangolo/fastapi/blob/master/fastapi/openapi/utils.py#L438), the `openapi_version` parameter is never used again after storing it in the [output dict](https://github.com/tiangolo/fastapi/blob/master/fastapi/openapi/utils.py#L465), leading to the described behavior. Additionally, the parameter is also never validated, so setting `openapi_version='asdf'` is silently accepted, but will - of course - produce an invalid OAS.

In my opinion, `fastapi.openapi.utils.get_openapi` should at least validate the version and probably also throw an exception if it cannot export an OAS of that version.

### Operating System

Linux

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

3.10.12

### Additional Context

Somewhat related to https://github.com/tiangolo/fastapi/discussions/10081

**Comment by SCSR-is-too-short-username** [comment, marked as answer]
From [fastapi/application.py:856](https://github.com/tiangolo/fastapi/blob/38f191dcd30c7e7b7d422b4406f8eb8c6805b09d/fastapi/applications.py#L856):

> The version string of OpenAPI.
> FastAPI will generate OpenAPI version 3.1.0, and will output that as
> the OpenAPI version. But some tools, even though they might be
> compatible with OpenAPI 3.1.0, might not recognize it as a valid.
> So you could override this value to trick those tools into using
> the generated OpenAPI. Have in mind that this is a hack. But if you
> avoid using features added in OpenAPI 3.1.0, it might work for your
> use case.
>
> This is not passed as a parameter to the `FastAPI` class to avoid
> giving the false idea that FastAPI would generate a different OpenAPI
> schema. It is only available as an attribute.

Have you tried if the tool you are using is able to process the openapi file with the version `3.0.x` even though the file is generated by the `3.1.0` spec? I don't know why the function that generates the openapi is implemented this way, but at least the quote gives a bit more info on why it doesn't work the way you expected.

**Reply by tringwald** [reply]
Unfortunately, that wouldn't work in my special case, as there are incompatibilities between `3.1.0` and `3.0.x` that the tool is unable to parse. 
Given your link, this behavior seems to be documented now as of yesterday's commit. Another comment could maybe be added directly to `fastapi.openapi.utils.get_openapi()`, but overall it should be clearer now that the exported OAS version is static and not affected by the `openapi_version` parameter.
